### PR TITLE
Time API

### PR DESCRIPTION
**Pull request by gonsie**
This is a new API which allows for user defined `tw_stime` types. It is implemented through `#define` macros, which should have minimal performance impact in the nominal case. Existing models do not need to change, unless they plan on supporting various `tw_stime` types.

See the previous conversation in #157. This PR has `develop` as the base. 

---

If this merge represents a feature addition to ROSS, the following items must be completed before the branch will be merged:

- [x] Document the feature on the blog (See the [website Contributing guide](https://github.com/ROSS-org/ross-org.github.io/blob/master/CONTRIBUTING.md)).
  Include a link to your blog post in the Pull Request.
- [x] Builds should cleanly compile with -Wall and -Wextra.
- [x] One or more TravisCI tests should be created (and they should pass)
- [x] Through the TravisCI tests, coverage should increase
- [x] Test with CODES to ensure everything continues to work

### CODE REVIEW

**Comment by gonsie** [comment]
So I was wrong, this does represent an API change to ROSS model. I've renamed `TWOPT_STIME` to `TWOPT_DOUBLE` to be explicit about the accepted values. 

This API change means that we'd have do an 8.0 release. Alternatively, we could re-add TWOPT_STIME (as a deprecated option) for a 7.2 release... removing TWOPT_STIME in an eventual 8.0 release.

**Comment by gonsie** [comment]
Since this is an API change, I've created a [CODES branch](https://xgitlab.cels.anl.gov/gonsie/codes/commits/ross-develop-159) for the update.

**Comment by gonsie** [comment]
There is no functionality change, so no new unit tests are needed. Yet, coverage has increased 😄

File: core/gvt/mpi_allreduce.c [hunk 1]
```diff
@@ -233,7 +233,7 @@ tw_gvt_step2(tw_pe *me)
 #endif
 
     if ((g_st_model_stats == GVT_STATS || g_st_model_stats == ALL_STATS) && g_tw_gvt_done % g_st_num_gvt == 0)
-        st_collect_model_data(me, (tw_stime)tw_clock_read() / g_tw_clock_rate, GVT_STATS);
+        st_collect_model_data(me, tw_clock_read() / g_tw_clock_rate, GVT_STATS);
```
Comment (caitlinross) [inline core/gvt/mpi_allreduce.c]: There should be a cast to double so `tw_clock_read() / g_tw_clock_rate` works correctly.

File: core/tw-sched.c [hunk 2]
```diff
@@ -262,7 +262,7 @@ static void tw_sched_batch(tw_pe * me) {
             if (g_st_engine_stats == RT_STATS || g_st_engine_stats == ALL_STATS)
                 st_collect_engine_data(me, RT_COL);
             if (g_st_model_stats == RT_STATS || g_st_model_stats == ALL_STATS)
-                st_collect_model_data(me, (tw_stime)current_rt / g_tw_clock_rate, RT_STATS);
+                st_collect_model_data(me, current_rt / g_tw_clock_rate, RT_STATS);
```
Comment (caitlinross) [inline core/tw-sched.c]: This needs a double cast as well.

File: core/tw-sched.c [hunk 3]
```diff
@@ -387,14 +387,14 @@ static void tw_sched_batch_realtime(tw_pe * me) {
...
-        if(g_st_rt_sampling && 
+        if(g_st_rt_sampling &&
                 tw_clock_read() - g_st_rt_samp_start_cycles > g_st_rt_interval)
         {
             tw_clock current_rt = tw_clock_read();
             if (g_st_engine_stats == RT_STATS || g_st_engine_stats == ALL_STATS)
                 st_collect_engine_data(me, RT_COL);
             if (g_st_model_stats == RT_STATS || g_st_model_stats == ALL_STATS)
-                st_collect_model_data(me, (tw_stime)current_rt / g_tw_clock_rate, RT_STATS);
+                st_collect_model_data(me, current_rt / g_tw_clock_rate, RT_STATS);
```
Comment (caitlinross) [inline core/tw-sched.c]: Double cast.

File: core/tw-sched.c [hunk 4]
```diff
@@ -491,12 +491,12 @@ void tw_scheduler_sequential(tw_pe * me) {
...
         tw_event_free(me, cev);
 
-        if(g_st_rt_sampling && 
+        if(g_st_rt_sampling &&
                 tw_clock_read() - g_st_rt_samp_start_cycles > g_st_rt_interval)
         {
             tw_clock current_rt = tw_clock_read();
             if (g_st_model_stats == RT_STATS || g_st_model_stats == ALL_STATS)
-                st_collect_model_data(me, (tw_stime)current_rt / g_tw_clock_rate, RT_STATS);
+                st_collect_model_data(me, current_rt / g_tw_clock_rate, RT_STATS);
```
Comment (caitlinross) [inline core/tw-sched.c]: Double cast.

File: core/tw-sched.c [hunk 5]
```diff
@@ -600,14 +600,14 @@ void tw_scheduler_conservative(tw_pe * me) {
...
-            if(g_st_rt_sampling && 
+            if(g_st_rt_sampling &&
                     tw_clock_read() - g_st_rt_samp_start_cycles > g_st_rt_interval)
             {
                 tw_clock current_rt = tw_clock_read();
                 if (g_st_engine_stats == RT_STATS || g_st_engine_stats == ALL_STATS)
                     st_collect_engine_data(me, RT_COL);
                 if (g_st_model_stats == RT_STATS || g_st_model_stats == ALL_STATS)
-                    st_collect_model_data(me, (tw_stime)current_rt / g_tw_clock_rate, RT_STATS);
+                    st_collect_model_data(me, current_rt / g_tw_clock_rate, RT_STATS);
```
Comment (caitlinross) [inline core/tw-sched.c]: double cast

**Comment by caitlinross** [comment]
I think once those few changes I pointed out are made, it's good to merge!

**Comment by caitlinross** [comment]
Oh yeah and @gonsie in regards to CODES, it's actually been [moved to GitHub](https://github.com/codes-org/codes) so you may want to submit the PR there instead of Gitlab.

**Comment by gonsie** [comment]
thanks @caitlinross! I've updated the PR with your change requests. Please double check that I'm doing the double casting in the way you want. Then merge, or give me the go-ahead.

thanks for the update on codes as well! I'll PR my update to them once we have a 7.2 release.

File: core/gvt/mpi_allreduce.c [hunk 6]
```diff
@@ -233,7 +233,7 @@ tw_gvt_step2(tw_pe *me)
 #endif
 
     if ((g_st_model_stats == GVT_STATS || g_st_model_stats == ALL_STATS) && g_tw_gvt_done % g_st_num_gvt == 0)
-        st_collect_model_data(me, tw_clock_read() / g_tw_clock_rate, GVT_STATS);
+        st_collect_model_data(me, (double)(tw_clock_read() / g_tw_clock_rate), GVT_STATS);
```
Comment (caitlinross) [inline core/gvt/mpi_allreduce.c]: Sorry I wasn't clear before. For all of these calls, it should be something like `(double) tw_clock_read() / g_tw_clock_rate`. Basically at least one needs to be converted to a double *before* the division takes place so everything after the decimal doesn't get truncated.

**Comment by gonsie** [comment]
Got it! 

For realz this time @caitlinross !

**Pull request merged**